src/dodal/common/beamlines/device_factory.py

```diff
@@ -0,0 +1,112 @@
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Generic, ParamSpec, TypeVar
+
+from bluesky.run_engine import call_in_bluesky_event_loop
+from ophyd_async.core import DEFAULT_TIMEOUT, Device
+
+from dodal.common.beamlines.beamline_utils import ACTIVE_DEVICES
+
+P = ParamSpec("P")
+T = TypeVar("T", bound=Device)
+_skip = bool | Callable[[], bool]
+
+
+@dataclass
+class DeviceInitializationConfig:
+    """
+    default_eager: connect or raise Exception at startup
+    default_set_name: use factory name as name of device
+    default_timeout: timeout for connecting to the device
+    default_mock: use Signals with mock backends for device
+    default_skip: mark the factory to be (conditionally) skipped
+    """
+
+    eager_connect: bool
+    use_factory_name: bool
+    timeout: float
+    mock: bool
+    skip: _skip
+
+
+class DeviceInitializationController(Generic[P, T]):
+    def __init__(self, config: DeviceInitializationConfig, factory: Callable[P, T]):
+        self._factory: Callable[P, T] = factory
+        self._config: DeviceInitializationConfig = config
+        self._cached_device: T | None = None
+
+    @property
+    def skip(self) -> bool:
+        return self._config.skip() if callable(self._config.skip) else self._config.skip
+
+    @property
+    def device(self) -> T | None:
+        return self._cached_device
+
+    def __call__(
+        self,
+        connect: bool | None = None,
+        name: str | None = None,
+        timeout: float | None = None,
+        mock: bool | None = None,
+        *args: P.args,
+        **kwargs: P.kwargs,
```

Comment:
Is there a usecase for passing args and kwargs to the factory function?

Reply:
It's currently used on i03 for the eiger and oav, i04, i04_1, i23 for the eiger.

I can definitely envisage other use cases when instantiating a device automatically with a default value but wanting to use other values when instantiating manually.

Reply:
can we resolve this conversation @coretl ?

Reply:
I agree we should. See my comment at https://github.com/DiamondLightSource/dodal/pull/597#pullrequestreview-2324687868. Why did this get resolved without being implemented?

Reply:
Discussed with @coretl offline. I need to think more about the `*args` and `**kwargs` usecase

Reply:
So please can we remove them until we get this usecase

Reply:
```python


 def __call__(
        self,
        connect: bool | None = None,
        name: str | None = None,
        timeout: float | None = None,
        mock: bool | None = None,
        *args: P.args,
        **kwargs: P.kwargs,
    ) -> T:
        # that is for the second and later times the device is called
        if self.device is not None:
            return self.device

        # unpack the arguments, fill in the defaults
        connect = connect or self._config.eager_connect
        name = name or (
            self._factory.__name__ if self._config.use_factory_name else None
        )
        mock = mock or self._config.mock
        timeout = timeout or self._config.timeout

        try:
            fake_with_ophyd_sim: object | None = kwargs.pop("fake_with_ophyd_sim")
            assert isinstance(
                fake_with_ophyd_sim, bool | None
            ), "fake_with_ophyd_sim must be a boolean"
            mock = fake_with_ophyd_sim or mock
            # the instantiation of the device
        except KeyError:
            pass

        device = self._factory(*args, **kwargs)
                if name:
            device.set_name(name)

        # connect the device if needed
        if connect:
            call_in_bluesky_event_loop(
                device.connect(
                    timeout=timeout,
                    mock=mock,
                )
            )

        self._cache_device(device)
        return device
 ```
 
 so the call signatre would change and the line `device = self._factory(*args, **kwargs)`
 
 becomes just `device = self._factory()`, right @coretl ?

Reply:
yes

Reply:
the linter then refuses to accept that.


> Discussed with @coretl offline. I need to think more about the *args and **kwargs usecase



could you please shed a bit more light on this?

Reply:
out of date